### ISSUE TITLE
Pop doesn't undo function declarations

### ISSUE THREAD

**Issue by ricardoV94** [egraphs-good/egglog#305]
I am not sure if it's supposed to, but the following fails:

```
(sort TP)
(push 1)
(function TP1 () TP)
(pop 1)
(push 1)
(function TP1 () TP)
(pop 1)
```

**Comment by yihozhang** [comment]
hmm, this seems to work for me? See [this demo](https://egraphs-good.github.io/egglog/?program=XQAAgAB8AAAAAAAAAAAUHMnnVi1HxWje8P80xC2Q7dUZMfokI8h91Ixbn4A72_8Xqg0lmeRQDI1BDuD6E222hwc30E6ZnxK4lF_PQXEd2Adthz2hxD__E-EAAA%253D%253D).

There's an issue that the outputs get replayed, but that seems to be a separate issue.

**Comment by ricardoV94** [comment]
CC @saulshanabrook

**Comment by saulshanabrook** [comment]
Yeah in the python bindings, I believe all the declarations (sorts and functions) are preserved between popped egraphs. I'll move this to that rep. 

It should be doable to fix it. If you have a particular use case you need this for that could be handy too to reference.

**Comment by ricardoV94** [comment]
I had a use case, just don't remember which. I'll let you know if it comes back
